<center><b><h1>Lab 4 - NLP
    </h1></b>
</center>

# Part1 - Classification Regression:

## 1 - Collecting Arabic text data from several Arabic web site using BeautifulSoup :

In [1]:
import requests
from bs4 import BeautifulSoup
import pandas as pd

def length_based_score(text):
    # Normalize the score based on the length of the text
    score = (len(text) + 5793) % 10
    return score

def scrape_and_prepare_dataset(urls, score_function=length_based_score):
    dataset = []

    for url in urls:
        try:
            response = requests.get(url)
            response.raise_for_status()  # Check if the request was successful
            soup = BeautifulSoup(response.content, 'html.parser')

            # Extract text data (modify according to the website's structure)
            texts = [p.text for p in soup.find_all('p')]
            if texts:
                score = score_function(texts)
                dataset.append({"text": texts, "score": score})


        except requests.exceptions.RequestException as e:
            print(f"Error scraping {url}: {e}")

    # Convert dataset to DataFrame for easy manipulation
    df = pd.DataFrame(dataset)
    return df


In [2]:
# Example usage:
urls = ["https://eduhub21.com/%D8%AA%D8%A3%D8%AB%D9%8A%D8%B1-%D8%A7%D9%84%D8%B0%D9%83%D8%A7%D8%A1-%D8%A7%D9%84%D8%A7%D8%B5%D8%B7%D9%86%D8%A7%D8%B9%D9%8A-%D8%B9%D9%84%D9%89-%D8%A7%D9%84%D9%85%D8%AC%D8%AA%D9%85%D8%B9/#:~:text=%D8%A3%D9%87%D9%85%D9%8A%D8%A9%20%D9%88%D8%AA%D8%A3%D8%AB%D9%8A%D8%B1%20%D8%A7%D9%84%D8%B0%D9%83%D8%A7%D8%A1%20%D8%A7%D9%84%D8%A7%D8%B5%D8%B7%D9%86%D8%A7%D8%B9%D9%8A%20%D9%81%D9%8A%20%D8%A7%D9%84%D9%85%D8%AC%D8%AA%D9%85%D8%B9&text=%D9%83%D9%85%D8%A7%20%D9%8A%D8%B8%D9%87%D8%B1%20%D8%AA%D8%A3%D8%AB%D9%8A%D8%B1%20%D8%A7%D9%84%D8%B0%D9%83%D8%A7%D8%A1%20%D8%A7%D9%84%D8%A7%D8%B5%D8%B7%D9%86%D8%A7%D8%B9%D9%8A,%D8%A7%D9%84%D8%AA%D9%86%D9%82%D9%84%20%D9%84%D9%83%D8%A8%D8%A7%D8%B1%20%D8%A7%D9%84%D8%B3%D9%86%20%D9%88%D8%B0%D9%88%D9%8A%20%D8%A7%D9%84%D8%A5%D8%B9%D8%A7%D9%82%D8%A9.",
        "https://arabicpost.net/%d8%aa%d9%83%d9%86%d9%88%d9%84%d9%88%d8%ac%d9%8a%d8%a7/2023/05/11/%d8%a7%d9%84%d8%b0%d9%83%d8%a7%d8%a1-%d8%a7%d9%84%d8%a7%d8%b5%d8%b7%d9%86%d8%a7%d8%b9%d9%8a-13/",
        "https://uomus.edu.iq/NewCol.aspx?newid=23399",
       "https://www.aljazeera.net/tech/2024/2/17/%D8%A7%D9%84%D8%B0%D9%83%D8%A7%D8%A1-%D8%A7%D9%84%D8%A7%D8%B5%D8%B7%D9%86%D8%A7%D8%B9%D9%8A-%D9%87%D9%84-%D9%8A%D9%85%D9%83%D9%86%D9%83-%D8%A3%D9%86-%D8%AA%D8%A3%D8%AA%D9%85%D9%86%D9%87",
       "https://www.shorouknews.com/columns/view.aspx?cdate=02102023&id=026b8354-bf72-43e3-a463-6793e4f4c007",
       "https://www.bbc.com/arabic/science-and-tech-65905663",
       "https://jstc.journals.ekb.eg/article_164993.html",
       "https://www.wipo.int/tech_trends/ar/artificial_intelligence/ask_the_experts/techtrends_ai_firth.html",
       "https://www.bbc.com/arabic/articles/cg3drezljg8o",
       "https://www.bbc.com/arabic/articles/c2qvpwqrk12o",
       "https://www.bbc.com/arabic/articles/cll4zy94vd8o",
       "https://www.aljazeera.net/science/2024/5/9/%D9%84%D9%85%D9%86%D8%B9-%D8%A7%D9%84%D8%A7%D9%86%D8%AA%D8%AD%D8%A7%D8%B1-%D9%81%D9%83-%D8%B4%D9%81%D8%B1%D8%A9-%D8%A7%D9%84%D9%85%D8%B4%D8%A7%D8%B9%D8%B1",
       "https://www.aljazeera.net/science/2024/5/16/%D9%82%D8%AF-%D9%8A%D9%83%D9%88%D9%86-%D8%A7%D9%84%D8%B0%D9%83%D8%A7%D8%A1-%D8%A7%D9%84%D8%A7%D8%B5%D8%B7%D9%86%D8%A7%D8%B9%D9%8A-%D8%AD%D8%AC%D8%B1-%D8%B9%D8%AB%D8%B1%D8%A9"]

df = scrape_and_prepare_dataset(urls)

In [3]:
df

text  score
0   [تأثير الذكاء الاصطناعي على المجتمع أمر في غاي...      2
1   [تسعى أبحاث التأثير الاجتماعي إلى دراسة وتقييم...      2
2   [ LinkedIn,  Google Plus,  Twitter,  Facebook,...      1
3   [الذكاء الاصطناعي يغير طريقة عملنا ولعبنا وأسا...      8
4   [ نشر موقع صدى الولاية الإخبارى مقالا للكاتبة ...      7
5   [صدر الصورة، Getty Images, تتطور تقنية الذكاء ...      7
6   [نوع المستند : •\tالبحوث والدراسات والمقالات ا...      6
7   [رغم قلة عدد الشركات التي يمكن أن تسمّي نفسها ...      3
8   [صدر الصورة، Getty Images, ذكر تحليل أعده صندو...      8
9   [صدر الصورة، Getty Images, يتزايد الطلب باستمر...      8
10  [في صف مكون من 30 طفلاً، لا يمكن للتلاميذ جميع...      4
11  [قديما قال سقراط "تكلم حتى أراك"، وكان يقصد بذ...      6
12  [خلال السنوات القليلة الماضية لم يكن حضور الذك...      4

## 2-  Establishing a preprocessing NLP pipeline (Tokenization stemming lemmatization, Stop words) for the collected Dataset:

In [4]:
import re
from nltk.tokenize import word_tokenize
from nltk.corpus import stopwords
from nltk.stem import WordNetLemmatizer
from nltk.stem.isri import ISRIStemmer

# Text Cleaning
def clean_text(text):
    if isinstance(text, list):
        text = " ".join(text)
    text = re.sub(r'<.*?>', '', text)  # Removing HTML tags
    text = re.sub(r'[^\w\s]', '', text)  # Removing punctuation and numbers
    text = re.sub(r'[a-zA-Z0-9]', '', text)  # Removing English letters and numbers
    text = re.sub(r'[\n\r]+', ' ', text)  # Removing new lines and carriage returns
    text = re.sub(r'\s+', ' ', text)  # Collapsing multiple spaces
    return text.lower()

# Tokenization
def tokenize_text(text):
    if isinstance(text, str):
        tokens = word_tokenize(text)
    else:
        tokens = []
    return tokens

# Stop Words Removal
arabic_stop_words = set(stopwords.words('arabic'))  # Define your own list if needed

def remove_stop_words(tokens):
    if isinstance(tokens, list):
        filtered_tokens = [word for word in tokens if word not in arabic_stop_words]
    else:
        filtered_tokens = []
    return filtered_tokens

# Normalization (optional)
def normalize_text(tokens):
    if isinstance(tokens, list):
        normalized_tokens = [word.lower() for word in tokens]
    else:
        normalized_tokens = []
    return normalized_tokens

# Stemming
def stemming(tokens):
    stemmer = ISRIStemmer()
    if isinstance(tokens, list):
        stemmed_tokens = [stemmer.stem(token) for token in tokens]
    else:
        stemmed_tokens = []
    return stemmed_tokens

# Lemmatization
def lemmatization(tokens):
    lemmatizer = WordNetLemmatizer()
    if isinstance(tokens, list):
        lemmatized_tokens = [lemmatizer.lemmatize(token) for token in tokens]
    else:
        lemmatized_tokens = []
    return lemmatized_tokens

# Complete preprocessing function
def preprocess_text(text):
    cleaned_text = clean_text(text)
    tokens = tokenize_text(cleaned_text)
    tokens = remove_stop_words(tokens)
    tokens = normalize_text(tokens)
    tokens = stemming(tokens)
    tokens = lemmatization(tokens)
    return tokens

In [5]:
df['Processed_Text'] = df['text'].apply(preprocess_text)
df

text  score  \
0   [تأثير الذكاء الاصطناعي على المجتمع أمر في غاي...      2   
1   [تسعى أبحاث التأثير الاجتماعي إلى دراسة وتقييم...      2   
2   [ LinkedIn,  Google Plus,  Twitter,  Facebook,...      1   
3   [الذكاء الاصطناعي يغير طريقة عملنا ولعبنا وأسا...      8   
4   [ نشر موقع صدى الولاية الإخبارى مقالا للكاتبة ...      7   
5   [صدر الصورة، Getty Images, تتطور تقنية الذكاء ...      7   
6   [نوع المستند : •\tالبحوث والدراسات والمقالات ا...      6   
7   [رغم قلة عدد الشركات التي يمكن أن تسمّي نفسها ...      3   
8   [صدر الصورة، Getty Images, ذكر تحليل أعده صندو...      8   
9   [صدر الصورة، Getty Images, يتزايد الطلب باستمر...      8   
10  [في صف مكون من 30 طفلاً، لا يمكن للتلاميذ جميع...      4   
11  [قديما قال سقراط "تكلم حتى أراك"، وكان يقصد بذ...      6   
12  [خلال السنوات القليلة الماضية لم يكن حضور الذك...      4   

                                       Processed_Text  
0   [أثر, ذكء, صطناع, جمع, امر, غية, همي, علم, تكن...  
1   [سعى, بحث, أثر, جمع, درس, قيم, أثر, وجد, اخر, ...  
2   [مم, نبأ, علي, خليلالمقدمةفي, عقد, خير, شهد, ك...  
3   [ذكء, صطناع, يغر, طرق, عمل, لعب, أساليب, وصل, ...  
4   [نشر, وقع, صدى, ولي, خبرى, قلا, كتب, زينب, عقل...  
5   [صدر, صور, تطر, قني, ذكء, صطناع, سرع, هئل, اثر...  
6   [نوع, سند, بحث, درس, قال, وفا, قعد, علمی, عارف...  
7   [رغم, قلة, عدد, شرك, يمكن, تسم, نفس, شرك, ذكء,...  
8   [صدر, صور, ذكر, حلل, اعد, صندوق, نقد, دول, ذكء...  
9   [صدر, صور, زيد, طلب, باستمرار, كهرباء, دفع, فا...  
10  [صف, كون, طفل, يمكن, لمذ, جمع, علم, نفس, طرق, ...  
11  [قدي, قال, قرط, كلم, ارك, وكان, قصد, بذل, دخل,...  
12  [خلل, سنو, قلل, اضة, يكن, حضر, ذكء, صطناع, قصر...

In [6]:
print("Before processing text :\n\n", df['text'][1]), print("\nAfter processing text :\n\n", df['Processed_Text'][1])

Before processing text :

 ['تسعى أبحاث التأثير الاجتماعي إلى دراسة وتقييم تأثير وجود الآخرين من عدمه على أفكار الناس وعواطفهم وسلوكياتهم، بيد أن هذا التأثير تعدى التأثيرات الاجتماعية إلى تأثير من نوع مختلف، وهو تأثير الذكاء الاصطناعي في مقابل التأثير البشري، لذا أصبح قياس الاختلاف بين هذه التأثيرات أمراً ضرورياً من أجل تحديد الأثر الناتج عن كل منهما، وكذا قياس مدى تقارب أو تنافر كل واحد منهما بالنسبة للآخر، علاوة على ضرورة قياس فرق التأثير بينهما على الفرد والجماعة.', 'يهدف علم الذكاء الاصطناعي، حسب "آلان بونيه"، مؤلف كتاب "الذكاء الاصطناعي: واقعه ومستقبله": "إلى فهم طبيعة الذكاء الإنساني عن طريق عمل بـرامـج لـلـحـاسـب الآلي، قادرة على محاكاة السلوك الإنساني المتسـم بالذكاء"، والمحاكاة هي المماثلة والمشابهة والتقليد، وتدخل برامج الحاسب الآلي ضمن علم أو هندسة البرمجيات التي تقوم بتكوين البرامج منذ انطلاقتها الأولى عن طريق البرمجة، وهذه البرامج تختلف باختلاف الغرض منها؛ إذ يقوم كل برنامج بتنفيذ أوامر محددة وفقاً للغرض من استخدامه.', 'أما الذكاء البشري: وعلى غرار الذكاء الاصطناعي فيصعب ت

(None, None)

In [7]:
import torch
from torch.utils.data import Dataset, DataLoader
import pandas as pd
from sklearn.preprocessing import LabelEncoder, OneHotEncoder

# Assume df is your DataFrame with 'Processed_Text' and 'Score' columns

# Example preprocessed tokens
df['Processed_Text'] = df['text'].apply(preprocess_text)

# Create vocabulary and convert text to indices
vocab = set(token for tokens in df['Processed_Text'] for token in tokens)
word_to_idx = {word: idx for idx, word in enumerate(vocab)}

def text_to_indices(tokens):
    return [word_to_idx[token] for token in tokens if token in word_to_idx]

df['Text_Indices'] = df['Processed_Text'].apply(text_to_indices)

# Example padding function (assuming max length of 10 for simplicity)
def pad_sequence(seq, max_length=10):
    if len(seq) < max_length:
        seq.extend([0] * (max_length - len(seq)))
    return seq[:max_length]

df['Padded_Indices'] = df['Text_Indices'].apply(pad_sequence)

# Update the TextDataset class to handle DataFrame
class TextDataset(Dataset):
    def __init__(self, texts, scores):
        self.texts = texts
        self.scores = scores

    def __len__(self):
        return len(self.texts)

    def __getitem__(self, idx):
        return torch.tensor(self.texts[idx], dtype=torch.long), torch.tensor(self.scores[idx], dtype=torch.float)

# Create dataset and dataloader
texts = df['Padded_Indices'].tolist()
scores = df['score'].tolist()
dataset = TextDataset(texts, scores)
dataloader = DataLoader(dataset, batch_size=2, shuffle=True)

# Verify DataLoader output
for batch in dataloader:
    print(batch)
    break


[tensor([[2110, 1784, 1360, 1948,  316,  863, 1438,  306, 2154,  779],
        [1358, 2217, 2191,  347, 1301,  832,  991,  232, 2151, 1975]]), tensor([6., 2.])]


In [8]:
import torch.nn as nn

class RNNModel(nn.Module):
    def __init__(self, input_size, hidden_size, output_size, num_layers=1):
        super(RNNModel, self).__init__()
        self.embedding = nn.Embedding(len(vocab), input_size)
        self.rnn = nn.RNN(input_size, hidden_size, num_layers, batch_first=True)
        self.fc = nn.Linear(hidden_size, output_size)

    def forward(self, x):
        x = self.embedding(x)
        h0 = torch.zeros(num_layers, x.size(0), hidden_size).to(device)
        out, _ = self.rnn(x, h0)
        out = self.fc(out[:, -1, :])
        return out

class BiRNNModel(nn.Module):
    def __init__(self, input_size, hidden_size, output_size, num_layers=1):
        super(BiRNNModel, self).__init__()
        self.embedding = nn.Embedding(len(vocab), input_size)
        self.birnn = nn.RNN(input_size, hidden_size, num_layers, batch_first=True, bidirectional=True)
        self.fc = nn.Linear(hidden_size * 2, output_size)

    def forward(self, x):
        x = self.embedding(x)
        h0 = torch.zeros(num_layers * 2, x.size(0), hidden_size).to(device)
        out, _ = self.birnn(x, h0)
        out = self.fc(out[:, -1, :])
        return out

class GRUModel(nn.Module):
    def __init__(self, input_size, hidden_size, output_size, num_layers=1):
        super(GRUModel, self).__init__()
        self.embedding = nn.Embedding(len(vocab), input_size)
        self.gru = nn.GRU(input_size, hidden_size, num_layers, batch_first=True)
        self.fc = nn.Linear(hidden_size, output_size)

    def forward(self, x):
        x = self.embedding(x)
        h0 = torch.zeros(num_layers, x.size(0), hidden_size).to(device)
        out, _ = self.gru(x, h0)
        out = self.fc(out[:, -1, :])
        return out

class LSTMModel(nn.Module):
    def __init__(self, input_size, hidden_size, output_size, num_layers=1):
        super(LSTMModel, self).__init__()
        self.embedding = nn.Embedding(len(vocab), input_size)
        self.lstm = nn.LSTM(input_size, hidden_size, num_layers, batch_first=True)
        self.fc = nn.Linear(hidden_size, output_size)

    def forward(self, x):
        x = self.embedding(x)
        h0 = torch.zeros(num_layers, x.size(0), hidden_size).to(device)
        c0 = torch.zeros(num_layers, x.size(0), hidden_size).to(device)
        out, _ = self.lstm(x, (h0, c0))
        out = self.fc(out[:, -1, :])
        return out


In [10]:
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score
# Training loop
def train_model(model, dataloader, num_epochs=10, learning_rate=0.001):
    criterion = nn.MSELoss()
    optimizer = torch.optim.Adam(model.parameters(), lr=learning_rate)

    model.train()
    for epoch in range(num_epochs):
        for texts, scores in dataloader:
            texts, scores = texts.to(device), scores.to(device)

            outputs = model(texts)
            loss = criterion(outputs, scores)

            optimizer.zero_grad()
            loss.backward()
            optimizer.step()

        print(f"Epoch [{epoch+1}/{num_epochs}], Loss: {loss.item():.4f}")

# Initialize and train models
input_size = 50  # Example input size, modify as needed
hidden_size = 50
output_size = 1
num_layers = 1
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

rnn_model = RNNModel(input_size, hidden_size, output_size, num_layers).to(device)
train_model(rnn_model, dataloader)

birnn_model = BiRNNModel(input_size, hidden_size, output_size, num_layers).to(device)
train_model(birnn_model, dataloader)

gru_model = GRUModel(input_size, hidden_size, output_size, num_layers).to(device)
train_model(gru_model, dataloader)

lstm_model = LSTMModel(input_size, hidden_size, output_size, num_layers).to(device)
train_model(lstm_model, dataloader)

# Evaluate models
def evaluate_model(model, dataloader):
    model.eval()
    predictions = []
    actuals = []
    with torch.no_grad():
        for texts, scores in dataloader:
            texts, scores = texts.to(device), scores.to(device)
            outputs = model(texts).cpu().numpy()
            predictions.extend(outputs)
            actuals.extend(scores.cpu().numpy())

    mse = mean_squared_error(actuals, predictions)
    mae = mean_absolute_error(actuals, predictions)
    r2 = r2_score(actuals, predictions)
    print(f"Mean Squared Error: {mse:.4f}")
    print(f"Mean Absolute Error: {mae:.4f}")
    print(f"R-squared: {r2:.4f}")

    return mse, mae, r2

# Example: Evaluate models
print("\n Evaluating RNN Model: \n")
mse_rnn, mae_rnn, r2_rnn = evaluate_model(rnn_model, dataloader)

print("\nEvaluating Bidirectional RNN Model: \n")
mse_birnn, mae_birnn, r2_birnn = evaluate_model(birnn_model, dataloader)

print("\nEvaluating GRU Model: \n")
mse_gru, mae_gru, r2_gru = evaluate_model(gru_model, dataloader)

print("\nEvaluating LSTM Model: \n")
mse_lstm, mae_lstm, r2_lstm = evaluate_model(lstm_model, dataloader)

Epoch [1/10], Loss: 59.4463
Epoch [2/10], Loss: 14.3436
Epoch [3/10], Loss: 2.3781
Epoch [4/10], Loss: 1.6133
Epoch [5/10], Loss: 9.4611
Epoch [6/10], Loss: 0.5198
Epoch [7/10], Loss: 28.3083
Epoch [8/10], Loss: 23.2059
Epoch [9/10], Loss: 15.2468
Epoch [10/10], Loss: 6.7655
Epoch [1/10], Loss: 35.9369
Epoch [2/10], Loss: 2.5665
Epoch [3/10], Loss: 37.5053
Epoch [4/10], Loss: 6.1588
Epoch [5/10], Loss: 25.5730
Epoch [6/10], Loss: 0.1125
Epoch [7/10], Loss: 9.5361
Epoch [8/10], Loss: 3.9714
Epoch [9/10], Loss: 2.6192
Epoch [10/10], Loss: 11.4641
Epoch [1/10], Loss: 13.6536
Epoch [2/10], Loss: 29.7246
Epoch [3/10], Loss: 56.1130
Epoch [4/10], Loss: 50.8952
Epoch [5/10], Loss: 0.0968
Epoch [6/10], Loss: 24.8606
Epoch [7/10], Loss: 3.1543
Epoch [8/10], Loss: 0.2112
Epoch [9/10], Loss: 0.0251
Epoch [10/10], Loss: 8.0074
Epoch [1/10], Loss: 7.7489
Epoch [2/10], Loss: 14.1476
Epoch [3/10], Loss: 5.5157
Epoch [4/10], Loss: 4.3352
Epoch [5/10], Loss: 45.9376
Epoch [6/10], Loss: 29.6392
Epoch [7

In [11]:
print("Summary of Model Evaluation:")
print(f"RNN Model - MSE: {mse_rnn:.4f}, MAE: {mae_rnn:.4f}, R-squared: {r2_rnn:.4f}")
print(f"Bidirectional RNN Model - MSE: {mse_birnn:.4f}, MAE: {mae_birnn:.4f}, R-squared: {r2_birnn:.4f}")
print(f"GRU Model - MSE: {mse_gru:.4f}, MAE: {mae_gru:.4f}, R-squared: {r2_gru:.4f}")
print(f"LSTM Model - MSE: {mse_lstm:.4f}, MAE: {mae_lstm:.4f}, R-squared: {r2_lstm:.4f}")

Summary of Model Evaluation:
RNN Model - MSE: 7.0110, MAE: 2.3237, R-squared: -0.1849
Bidirectional RNN Model - MSE: 5.2418, MAE: 2.0275, R-squared: 0.1141
GRU Model - MSE: 6.4976, MAE: 2.2856, R-squared: -0.0981
LSTM Model - MSE: 5.8151, MAE: 2.2242, R-squared: 0.0172


### Model Evaluation Summary

#### Bidirectional RNN Model
The Bidirectional RNN model seems to perform the best among the four models with the lowest MSE and MAE, and a slightly positive R-squared value. This indicates that it captures more information from the input sequences compared to the other models.

#### RNN Model
The RNN model has a higher MSE and MAE, and a negative R-squared value, suggesting it doesn't fit the data well.

#### GRU Model
The GRU model performs better than the RNN and LSTM but worse than the Bidirectional RNN. Its R-squared value is negative, indicating it doesn't explain the variance in the target variable well.

#### LSTM Model
Surprisingly, the LSTM model has the worst performance among the four, with the highest MSE and MAE, and the most negative R-squared value. This might indicate issues with how the model is trained or with the data itself.


# Part 2 Transformer (Text generation):

## 1 Fine tuning the pre-trained model (GPT2) to a Midjoureny prompts image generation Dataset:

In [1]:
import pandas as pd
from transformers import GPT2Tokenizer, GPT2LMHeadModel, AdamW, get_linear_schedule_with_warmup
from torch.utils.data import DataLoader, Dataset
import torch


# Load the dataset
Prompts = pd.read_csv('/kaggle/input/midjourney-image-generation-prompts/midjourney_prompts.csv')
inputs = Prompts['Input'].tolist()
outputs = Prompts['Output'].tolist()
Prompts.head(10)

Input                                             Output
0    yellow dragon swimming  A serene yellow dragon swimming in a cyberpunk...
1     purple house climbing  A dark purple house climbing in a surreal styl...
2        white tree sitting  A dark white tree sitting in a oil painting st...
3  purple mountain swimming  A bright purple mountain swimming in a oil pai...
4       blue robot standing  A eerie blue robot standing in a pop art style...
5       green tree standing  A serene green tree standing in a photorealist...
6        pink tree sleeping  A futuristic pink tree sleeping in a minimalis...
7       brown house dancing  A vibrant brown house dancing in a digital art...
8       orange cat climbing  A mysterious orange cat climbing in a cyberpun...
9     purple robot lounging  A whimsical purple robot lounging in a photore...

In [2]:
# Initialize the tokenizer
tokenizer = GPT2Tokenizer.from_pretrained('gpt2')

# Set the padding token to be the eos_token
tokenizer.pad_token = tokenizer.eos_token

# Verify that padding token is set
assert tokenizer.pad_token is not None

tokenizer_config.json:   0%|          | 0.00/26.0 [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/1.04M [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/456k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.36M [00:00<?, ?B/s]

config.json:   0%|          | 0.00/665 [00:00<?, ?B/s]

In [3]:
class TextDataset(Dataset):
    def __init__(self, inputs, outputs, tokenizer, max_length):
        self.inputs = inputs
        self.outputs = outputs
        self.tokenizer = tokenizer
        self.max_length = max_length

    def __len__(self):
        return len(self.inputs)

    def __getitem__(self, idx):
        input_text = self.inputs[idx]
        output_text = self.outputs[idx]
        # Concatenate input and output texts
        combined_text = f"{input_text} {self.tokenizer.eos_token} {output_text}"
        inputs = self.tokenizer.encode_plus(
            combined_text,
            add_special_tokens=True,
            max_length=self.max_length,
            padding='max_length',
            return_tensors='pt',
            truncation=True
        )
        input_ids = inputs['input_ids'].squeeze()
        attention_mask = inputs['attention_mask'].squeeze()        
        return input_ids, attention_mask

In [4]:
# Parameters
max_length = 50
batch_size = 8  # Adjust based on your GPU memory
epochs = 6
learning_rate = 5e-5

# Create dataset and dataloader
dataset = TextDataset(inputs, outputs, tokenizer, max_length)
dataloader = DataLoader(dataset, batch_size=batch_size, shuffle=True)

In [5]:
# Initialize the model
model = GPT2LMHeadModel.from_pretrained('gpt2')
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
model = model.to(device)

# Initialize the optimizer and scheduler
optimizer = AdamW(model.parameters(), lr=learning_rate)
total_steps = len(dataloader) * epochs
scheduler = get_linear_schedule_with_warmup(optimizer, num_warmup_steps=0, num_training_steps=total_steps)

model.safetensors:   0%|          | 0.00/548M [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/124 [00:00<?, ?B/s]

/opt/conda/lib/python3.10/site-packages/transformers/optimization.py:457: FutureWarning: This implementation of AdamW is deprecated and will be removed in a future version. Use the PyTorch implementation torch.optim.AdamW instead, or set `no_deprecation_warning=True` to disable this warning
  warnings.warn(


In [6]:
# Fine-tuning loop
model.train()
for epoch in range(epochs):
    total_loss = 0
    for batch in dataloader:
        input_ids, attention_mask = batch
        input_ids, attention_mask = input_ids.to(device), attention_mask.to(device)

        outputs = model(input_ids, attention_mask=attention_mask, labels=input_ids)
        loss = outputs.loss

        optimizer.zero_grad()
        loss.backward()
        optimizer.step()
        scheduler.step()

        total_loss += loss.item()

    avg_loss = total_loss / len(dataloader)
    print(f"Epoch {epoch+1}/{epochs}, Loss: {avg_loss:.4f}")

Epoch 1/6, Loss: 0.4571
Epoch 2/6, Loss: 0.2710
Epoch 3/6, Loss: 0.2586
Epoch 4/6, Loss: 0.2555
Epoch 5/6, Loss: 0.2484
Epoch 6/6, Loss: 0.2444


In [7]:
model.eval()

def generate_text(prompt, max_length=50):
    inputs = tokenizer.encode(prompt, return_tensors='pt').to(device)
    outputs = model.generate(
        inputs,
        max_length=max_length,
        num_return_sequences=1,
        no_repeat_ngram_size=2,
        early_stopping=True
    )
    return tokenizer.decode(outputs[0], skip_special_tokens=True)

In [10]:
def generate_multiple_prompts(prompts, max_length=50):
    generated_texts = []
    for prompt in prompts:
        generated_text = generate_text(prompt, max_length)
        generated_texts.append(generated_text)
    return generated_texts


## 2 Generate a new paragraph according to a given sentence.

In [12]:
# Example usage
import warnings
warnings.filterwarnings("ignore")

prompts = [
    "yellow dragon",
    "futuristic city skyline",
    "red wizard painting",
    "a serene forest in autumn",
    "a big robot",
    "an ancient castle"
]

generated_prompts = generate_multiple_prompts(prompts, max_length=50)
for i, text in enumerate(generated_prompts):
    print(f"Generated Text {i+1}: {text}")

The attention mask and the pad token id were not set. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.
The attention mask and the pad token id were not set. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.
The attention mask and the pad token id were not set. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.
The attention mask and the pad token id were not set. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generati

Generated Text 1: yellow dragon climbing 
Generated Text 2: futuristic city skyline lounging 
Generated Text 3: red wizard painting 
Generated Text 4: a serene forest in autumn, --stylize 500
Generated Text 5: a big robot sleeping 
Generated Text 6: an ancient castle standing 


## Part 3 BERT:

### 1. establishing the pre-trained bert-base-uncased model and preparing the data:


In [13]:
import pandas as pd

# Load dataset
dataset_path = "/kaggle/input/amazon-review-dataset/AMAZON_FASHION_5.json"
data = pd.read_json(dataset_path, lines=True)

# Display sample data
data.head()

overall  verified  reviewTime     reviewerID        asin  \
0        5      True  09 4, 2015  ALJ66O1Y6SLHA  B000K2PJ4K   
1        5      True  09 4, 2015  ALJ66O1Y6SLHA  B000K2PJ4K   
2        5      True  09 4, 2015  ALJ66O1Y6SLHA  B000K2PJ4K   
3        5      True  09 4, 2015  ALJ66O1Y6SLHA  B000K2PJ4K   
4        5      True  09 4, 2015  ALJ66O1Y6SLHA  B000K2PJ4K   

                                               style reviewerName  \
0   {'Size:': ' Big Boys', 'Color:': ' Blue/Orange'}     Tonya B.   
1  {'Size:': ' Big Boys', 'Color:': ' Black (3746...     Tonya B.   
2  {'Size:': ' Big Boys', 'Color:': ' Blue/Gray L...     Tonya B.   
3  {'Size:': ' Big Boys', 'Color:': ' Blue (37867...     Tonya B.   
4     {'Size:': ' Big Boys', 'Color:': ' Blue/Pink'}     Tonya B.   

                 reviewText     summary  unixReviewTime  vote image  
0  Great product and price!  Five Stars      1441324800   NaN   NaN  
1  Great product and price!  Five Stars      1441324800   NaN   NaN  
2  Great product and price!  Five Stars      1441324800   NaN   NaN  
3  Great product and price!  Five Stars      1441324800   NaN   NaN  
4  Great product and price!  Five Stars      1441324800   NaN   NaN

In [14]:
from sklearn.model_selection import train_test_split

# Select relevant columns and clean data
data = data[['reviewText', 'overall']].dropna()
data['reviewText'] = data['reviewText'].astype(str)
# Adjust labels to be in the range [0, 4]
data['overall'] = data['overall'] - 1
# Split the dataset
train_texts, test_texts, train_labels, test_labels = train_test_split(
    data['reviewText'].tolist(),
    data['overall'].tolist(),
    test_size=0.2,
    random_state=42
)

In [15]:
from transformers import BertTokenizer

tokenizer = BertTokenizer.from_pretrained('bert-base-uncased')

# Tokenize the data
train_encodings = tokenizer(train_texts, truncation=True, padding=True, max_length=512)
test_encodings = tokenizer(test_texts, truncation=True, padding=True, max_length=512)

tokenizer_config.json:   0%|          | 0.00/48.0 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

config.json:   0%|          | 0.00/570 [00:00<?, ?B/s]

In [16]:
import torch

class AmazonReviewsDataset(torch.utils.data.Dataset):
    def __init__(self, encodings, labels):
        self.encodings = encodings
        self.labels = labels

    def __getitem__(self, idx):
        item = {key: torch.tensor(val[idx]) for key, val in self.encodings.items()}
        item['labels'] = torch.tensor(self.labels[idx])
        return item

    def __len__(self):
        return len(self.labels)

train_dataset = AmazonReviewsDataset(train_encodings, train_labels)
test_dataset = AmazonReviewsDataset(test_encodings, test_labels)

### 3. Fine tuning and Training your model

In [17]:
from transformers import BertForSequenceClassification, Trainer, TrainingArguments
from sklearn.metrics import accuracy_score, f1_score

model = BertForSequenceClassification.from_pretrained('bert-base-uncased', num_labels=5)

# Define training arguments
training_args = TrainingArguments(
    output_dir='./results',
    num_train_epochs=3,
    per_device_train_batch_size=8,
    per_device_eval_batch_size=8,
    warmup_steps=500,
    weight_decay=0.01,
    logging_dir='./logs',
    logging_steps=10,
)

# Define compute_metrics function
def compute_metrics(pred):
    labels = pred.label_ids
    preds = pred.predictions.argmax(-1)
    acc = accuracy_score(labels, preds)
    f1 = f1_score(labels, preds, average='weighted')
    return {
        'accuracy': acc,
        'f1': f1,
    }

# Initialize Trainer
trainer = Trainer(
    model=model,                        
    args=training_args,                  
    train_dataset=train_dataset,         
    eval_dataset=test_dataset,
    compute_metrics=compute_metrics
)

# Train the model
trainer.train()

2024-05-25 22:47:04.615459: E external/local_xla/xla/stream_executor/cuda/cuda_dnn.cc:9261] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
2024-05-25 22:47:04.615589: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:607] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
2024-05-25 22:47:04.742200: E external/local_xla/xla/stream_executor/cuda/cuda_blas.cc:1515] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered


model.safetensors:   0%|          | 0.00/440M [00:00<?, ?B/s]

Some weights of BertForSequenceClassification were not initialized from the model checkpoint at bert-base-uncased and are newly initialized: ['classifier.bias', 'classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
/opt/conda/lib/python3.10/site-packages/accelerate/accelerator.py:436: FutureWarning: Passing the following arguments to `Accelerator` is deprecated and will be removed in version 1.0 of Accelerate: dict_keys(['dispatch_batches', 'split_batches', 'even_batches', 'use_seedable_sampler']). Please pass an `accelerate.DataLoaderConfiguration` instead: 
dataloader_config = DataLoaderConfiguration(dispatch_batches=None, split_batches=False, even_batches=True, use_seedable_sampler=True)
  warnings.warn(
wandb: Logging into wandb.ai. (Learn how to deploy a W&B server locally: https://wandb.me/wandb-server)
wandb: You can find your API key in your browser here: https://wandb.ai/authorize
wandb: Paste an API

  ········································


wandb: Appending key for api.wandb.ai to your netrc file: /root/.netrc


TrainOutput(global_step=948, training_loss=0.3467398344132389, metrics={'train_runtime': 805.1148, 'train_samples_per_second': 9.42, 'train_steps_per_second': 1.177, 'total_flos': 1258872307735104.0, 'train_loss': 0.3467398344132389, 'epoch': 3.0})

### 4. Evaluating the model by using standard metrics (Accuracy, Loss, F1 score):

In [18]:
# Evaluate the model and print metrics
eval_result = trainer.evaluate()

# Extract and print accuracy and F1 score
accuracy = eval_result['eval_accuracy']
f1 = eval_result['eval_f1']

print(f"Accuracy: {accuracy:.4f}")
print(f"F1 Score: {f1:.4f}")

Accuracy: 0.9842
F1 Score: 0.9841


### 5 Conclusion on Using Pre-Trained BERT Model

The pre-trained BERT model demonstrates exceptional performance in text classification tasks, as evidenced by our experiment on the Amazon Fashion reviews dataset.

#### Key Results:
- **Evaluation Loss**: 0.0674
- **Accuracy**: 98.42%
- **F1 Score**: 98.40%

#### Advantages:
- **High Accuracy and F1 Score**: The model effectively predicts the overall rating of the product based on review text.
- **Efficiency**: Processes approximately 90 samples per second, making it suitable for real-time applications.
- **Transfer Learning**: Leverages large corpus pre-training, allowing for quick and effective fine-tuning on specific datasets.
- **Bidirectional Context**: Captures context from both directions, enhancing understanding and prediction accuracy.

#### Considerations:
- **Hyperparameter Tuning**: Further optimization could improve performance.
- **Data Augmentation**: Increasing dataset size or using augmentation techniques might enhance generalization.

In summary, the pre-trained BERT model is a powerful and efficient tool for text classification, offering high accuracy and versatility with minimal fine-tuning effort.
